pkg/apis/rollouts/validation/validation_test.go

```diff
@@ -0,0 +1,190 @@
+}
+
+func TestValidateRolloutStrategy(t *testing.T) {
+	rollout := v1alpha1.Rollout{
+		Spec: v1alpha1.RolloutSpec{
+			Strategy: v1alpha1.RolloutStrategy{},
+		},
+	}
+
+	allErrs := ValidateRolloutStrategy(&rollout, field.NewPath(""))
+	message := fmt.Sprintf(MissingFieldMessage, ".Spec.Strategy.Canary or .Spec.Strategy.BlueGreen")
+	assert.Equal(t, message, allErrs[0].Detail)
+
+	rollout.Spec.Strategy.BlueGreen = &v1alpha1.BlueGreenStrategy{}
+	rollout.Spec.Strategy.Canary = &v1alpha1.CanaryStrategy{}
+	allErrs = ValidateRolloutStrategy(&rollout, field.NewPath(""))
+	assert.Equal(t, InvalidStrategyMessage, allErrs[0].Detail)
+}
+
+func TestValidateRolloutStrategyBlueGreen(t *testing.T) {
+	scaleDownDelayRevisionLimit := defaults.DefaultRevisionHistoryLimit + 1
+	autoPromotionSeconds := int32(30)
+	rollout := v1alpha1.Rollout{
+		Spec: v1alpha1.RolloutSpec{
+			Strategy: v1alpha1.RolloutStrategy{
+				BlueGreen: &v1alpha1.BlueGreenStrategy{
+					PreviewService:              "service-name",
+					ActiveService:               "service-name",
+					ScaleDownDelayRevisionLimit: &scaleDownDelayRevisionLimit,
+					AutoPromotionSeconds:        &autoPromotionSeconds,
+				},
+			},
+		},
+	}
+
+	allErrs := ValidateRolloutStrategyBlueGreen(&rollout, field.NewPath("spec", "strategy", "blueGreen"))
+	assert.Len(t, allErrs, 3)
+	assert.Equal(t, DuplicatedServicesBlueGreenMessage, allErrs[0].BadValue)
+	assert.Equal(t, ScaleDownLimitLargerThanRevisionLimit, allErrs[1].Detail)
+	assert.Equal(t, InvalidAutoPromotionSecondsMessage, allErrs[2].Detail)
+}
+
+func TestValidateRolloutStrategyCanary(t *testing.T) {
+	setWeight := int32(101)
+	pauseDuration := intstr.FromInt(-1)
+	canaryStrategy := &v1alpha1.CanaryStrategy{
+		CanaryService: "stable-service",
+		StableService: "stable-service",
+		TrafficRouting: &v1alpha1.RolloutTrafficRouting{
+			SMI: &v1alpha1.SMITrafficRouting{},
+		},
+		Steps: []v1alpha1.CanaryStep{{}},
+	}
+	ro := &v1alpha1.Rollout{}
+	ro.Spec.Strategy.Canary = canaryStrategy
+
+	allErrs := ValidateRolloutStrategyCanary(ro, field.NewPath(""))
+	assert.Equal(t, DuplicatedServicesCanaryMessage, allErrs[0].Detail)
+
+	ro.Spec.Strategy.Canary.CanaryService = ""
+	allErrs = ValidateRolloutStrategyCanary(ro, field.NewPath(""))
+	assert.Equal(t, InvalidTrafficRoutingMessage, allErrs[0].Detail)
+
+	ro.Spec.Strategy.Canary.CanaryService = "canary-service"
+	allErrs = ValidateRolloutStrategyCanary(ro, field.NewPath(""))
+	assert.Equal(t, InvalidStepMessage, allErrs[0].Detail)
+
+	ro.Spec.Strategy.Canary.Steps[0].SetWeight = &setWeight
+	allErrs = ValidateRolloutStrategyCanary(ro, field.NewPath(""))
+	assert.Equal(t, InvalidSetWeightMessage, allErrs[0].Detail)
+
+	ro.Spec.Strategy.Canary.Steps[0].Pause = &v1alpha1.RolloutPause{
+		Duration: &pauseDuration,
+	}
+	ro.Spec.Strategy.Canary.Steps[0].SetWeight = nil
+	allErrs = ValidateRolloutStrategyCanary(ro, field.NewPath(""))
+	assert.Equal(t, InvalidDurationMessage, allErrs[0].Detail)
```

Comment:
I think the readability of this test can be improved by using t.Run to describe what check you are running and for each t.Run make a copy of the original rollout and modify the copied rollout. This helps isolate the tests because we can easily see what changes are required to the rollout for the test. It's definitely more verbose, but it helps future readers.

Reply:
For example, this block (line 127 to 132) can be rewritten as the following:
```golang
t.Run("invalid duration set in paused step", func(t *testing.T) {
	pauseDuration := intstr.FromInt(-1)
        invalidRo := ro.DeepCopy()
  	invalidRo.Spec.Strategy.Canary.Steps[0].Pause = &v1alpha1.RolloutPause{
		Duration: &pauseDuration,
	}
	allErrs = ValidateRolloutStrategyCanary(invalidRo, field.NewPath(""))
	assert.Equal(t, InvalidDurationMessage, allErrs[0].Detail)
})
```